Cluster features on raft: new procedure for joining group 0

piodul opened the pull request:
This PR implements a new procedure for joining nodes to group 0, based on the description in the "Cluster features on Raft (v2)" document. This is a continuation of the previous PRs related to cluster features on raft (https://github.com/scylladb/scylladb/pull/14722, https://github.com/scylladb/scylladb/pull/14232), and the last piece necessary to replace cluster feature checks in gossip.

Current implementation relies on gossip shadow round to fetch the set of enabled features, determine whether the node supports all of the enabled features, and joins only if it is safe. As we are moving management of cluster features to group 0, we encounter a problem: the contents of group 0 itself may depend on features, hence it is not safe to join it unless we perform the feature check which depends on information in group 0. Hence, we have a dependency cycle.

In order to solve this problem, the algorithm for joining group 0 is modified, and verification of features and other parameters is offloaded to an existing node in group 0. Instead of directly asking the discovery leader to unconditionally add the node to the configuration with `GROUP0_MODIFY_CONFIG`, two different RPCs are added: `JOIN_NODE_REQUEST` and `JOIN_NODE_RESPONSE`. The main idea is as follows:

- The new node sends `JOIN_NODE_REQUEST` to the discovery leader. It sends a bunch of information describing the node, including supported cluster features. The discovery leader verifies some of the parameters and adds the node in the `none` state to `system.topology`.
- The topology coordinator picks up the request for the node to be joined (i.e. the node in `none` state), verifies its properties - including cluster features - and then:
	- If the node is accepted, the coordinator transitions it to `boostrap`/`replace` state and transitions the topology to `join_group0` state. The node is added to group 0 and then `JOIN_NODE_RESPONSE` is sent to it with information that the node was accepted.
	- Otherwise, the node is moved to `left` state, told by the coordinator via `JOIN_NODE_RESPONSE` that it was rejected and it shuts down.

The procedure is not retryable - if a node fails to do it from start to end and crashes in between, it will not be allowed to retry it with the same host_id - `JOIN_NODE_REQUEST` will fail. The data directory must be cleared before attempting to add it again (so that a new host_id is generated).

More details about the procedure and the RPC are described in `topology-over-raft.md`.

Fixes: #15152

piodul commented:
Some nice-to-haves that I didn't implement yet, but perhaps could be done in followups:

- Error injection tests to verify idempotence properties of the procedure
- Maybe add a new key in `system.scylla_local` to store the enabled features that were sent by the topology coordinator on boot instead of reusing the existing and legacy `enabled_features` key

gleb-cloudius commented on docs/dev/topology-over-raft.md:
```diff
@@ -216,3 +216,94 @@ There are also a few static columns for cluster-global properties:
+  or rejected.
+
+In order to correlate requests with replies, the joining node uses a UUID called
+"join cookie". Until the joining node acknowledges a reply from the coordinator,
```
I did not look at the code how this "join cookie" is used, but can host_id can be used instead?

kbr-scylla replied:
I'm guessing that host_id is generated and persisted early in the procedure. If the procedure crashes after persisting host_id, the next attempt might use some different parameters (which ones though?) but a different host_id.

gleb-cloudius replied:
If "host_id is generated and persisted early in the procedure" then after the crash the same host_id will be used. Theoretically one may change some parameters (cluster name? Snitch?) and and restart, but if we persist those before start joining we can detect the change and fail early. If features are changed we can detect it in join_request rpc since whatever was added into the topology will not match.

kostja replied:
One way to deal with lack of good names for the verb is to rename group0_modify_config to group0_modify_config_legacy and keep using the name group0_modify_config

piodul replied:
> One way to deal with lack of good names for the verb is to rename group0_modify_config to group0_modify_config_legacy and keep using the name group0_modify_config

I suppose that it's not a problem that the `join_node` name is less clear than `group0_modify_config`; the bigger problem is that we have two RPCs now which are both used in the handshake named `join_node_request` and `join_node_response`, and _both_ of them are two-way. It's confusing to talk about "response to join_node_request" etc.

gleb-cloudius commented on docs/dev/topology-over-raft.md:
```diff
@@ -216,3 +216,94 @@ There are also a few static columns for cluster-global properties:
+    from the `system.topology` table.
+
+In case of other failures (timeouts, connection issues etc.) the request is
+simply retried.
```
If a node that requests joining is permanently dead we need to stop retrying eventually.

piodul replied:
Yes. This should be implemented as a part of the error handling work.

gleb-cloudius commented on service/join_node.hh:
```diff
@@ -0,0 +1,87 @@
+/*
```
Consider moving to `service/raft/`

piodul replied:
Done

gleb-cloudius commented on service/raft/group0_state_machine.cc:
```diff
@@ -26,6 +26,7 @@
 #include "idl/frozen_schema.dist.impl.hh"
 #include "idl/experimental/broadcast_tables_lang.dist.hh"
 #include "idl/experimental/broadcast_tables_lang.dist.impl.hh"
+#include "service/join_node.hh"
```
Why do you need to add this here and in the files below in this patch?

piodul replied:
They aren't needed. I'll remove them. Perhaps I needed them in some previous version of the implementation, but I don't remember why.

gleb-cloudius commented on service/raft/raft_group0.cc:
```diff
@@ -461,6 +463,7 @@ future<> raft_group0::join_group0(std::vector<gms::inet_address> seeds, bool as_
                 // In a fresh cluster this will trigger an empty snapshot transfer which is redundant but correct.
                 // See #14066.
                 nontrivial_snapshot = true;
+                is_discovery_leader = true;
```
Can `is_discovery_leader` be unified with `nontrivial_snapshot`?

piodul replied:
I agree with Kamil, and that was the reason I kept them separate.

gleb-cloudius replied:
IMO if we will only leave  `is_discovery_leader` it will even clarify the code. It will be clear that only if a node is discovery_leader nontrivial snapshot is required.

piodul replied:
This is no longer relevant in v2, as the bool was removed.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -6038,6 +6077,187 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+            if (it == _topology_state_machine._topology.normal_nodes.end()) {
+                result.result = join_node_request_result::rejected {
+                    .reason = ::format("Cannot replace node {} because it is not in the 'normal' state", *params.replaced_id),
+                };
```
This is pre-existing but the check should be in the coordinator before is starts executing the command since by then the state of the node may change.

piodul replied:
Forgot to fix this before, but now this check is moved to the topology coordinator (`validate_joining_node`).

kbr-scylla commented:
@gusev-p please review as well. Understanding this new procedure will probably be important to implement upgrades.

I'm guessing that to correctly handle potential node bootstraps in a mixed cluster (where some nodes have upgraded to use raft-topology but others did not), we'll want to distinguish whether the joining node uses the new RPC or the old GROUP0_MODIFY_CONFIG. And probably in a fully upgraded cluster we'll reject attempts that use the old way.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -6038,6 +6077,187 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+                if (params.join_cookie != *rs.join_cookie) {
+                    // The topology coordinator still has to process the old request
+                    // and the node should reject it, only then the new request
+                    // can be placed
```
I do not understand what old request and why do we need all this dense with cookies and retries. What exactly the scenario that you handles here? Why would a not send more then one join request? Because it restarts? Why is it not enough to check that the parameters for the new request are exactly the same as the previous one and return an error otherwise? There should be only one request for a given host_id.

gleb-cloudius commented on db/system_keyspace.cc:
```diff
@@ -2676,8 +2676,13 @@ service::topology_features system_keyspace::decode_topology_features_state(::sha
+            auto features = decode_features(deserialize_set_column(*topology(), row, "supported_features"));
+            if (nstate == service::node_state::normal) {
+                ret.normal_supported_features.emplace(host_id, std::move(features));
+            } else if (nstate == service::node_state::none || nstate == service::node_state::replacing) {
```
Now I see how it is used. We do not store features in the regular node maps, but have this structure instead. As far as I understand this is for early loading. But OTOH we need to only early load features for normal nodes. May be we can have a separate structure for early check, but also store features together with all other node's metadata as well.

piodul replied:
I agree 100%. Supported features used to be a field of `replica_state` before, but I moved them to a separate struct in https://github.com/scylladb/scylladb/pull/14955 because they were needed to be loaded early. However, I agree that both having this struct and `supported_features` field in `replica_state` as you propose is more convenient. I refactored the code in this way in v2.

gleb-cloudius commented on service/topology_state_machine.hh:
```diff
@@ -87,6 +87,9 @@ struct topology_features {
     // Supported features, for normal nodes
     std::unordered_map<raft::server_id, std::set<sstring>> normal_supported_features;
 
+    // Supported features, for new nodes
```
Nodes that are waiting to be joined describe it better then "new" I think.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3300/

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -6038,6 +6077,187 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+                slogger.info("raft topology: join: returning ok for idempotency ({})", params.host_id);
+                co_return result;
+            } else {
+                assert(rs.state == node_state::none);
```
Please use `on_internal_error` instead

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -6038,6 +6077,187 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+    assert(this_shard_id() == 0);
+
+    // Serialize handling the responses, just to be sure
+    auto lock = co_await get_units(_join_node_response_handler_mutex, 1);
```
IIUC it could happen that we handle this message before we finish the two-way REQUEST RPC; is that OK?

piodul replied:
Should not happen now, the join_node_response handler now waits until join_node_request was sent and finished successfully.

piodul commented on docs/dev/topology-over-raft.md:
```diff
@@ -216,3 +216,94 @@ There are also a few static columns for cluster-global properties:
+On receiving `JOIN_NODE_REQUEST`, the node verifies some parameters included
+in request. If the verification fails, a response is immediately sent with
+information about rejection. If the verification is successful, the node
+looks at the `system.topology` under a group 0 guard:
```
In the current implementation, verification done by the node receiving JOIN_NODE_REQUEST checks the cluster and snitch names. AFAIK those are supposed to be cluster-wide constants, so no group 0 guard is necessary. Those parameters also aren't written to `system.topology`, so it makes more sense for JOIN_NODE_REQUEST recipient to do this verification.

piodul commented on docs/dev/topology-over-raft.md:
```diff
@@ -216,3 +216,94 @@ There are also a few static columns for cluster-global properties:
+
+Upon noticing a node in `none` state and with `join` topology request, the
+coordinator verifies parameters of the node (e.g. whether it supports all
+of the enabled features) and decides whether to accept it or not. It sends
```
I explained the need for initial verification here: https://github.com/scylladb/scylladb/pull/15196#discussion_r1309780775

As for the topology coordinator, we use it to do the "ultimate" verification and marking the node as accepted (i.e. continuing with bootstrapping/replacing the node) because we need to serialize it with marking a feature as enabled, which is currently also done by the topology coordinator.

piodul commented on docs/dev/topology-over-raft.md:
```diff
@@ -216,3 +216,94 @@ There are also a few static columns for cluster-global properties:
+  removed from the `system.topology` table.
+- If the joining node accepts the decision and responds successfully:
+  - If the decision was to add the node, then it is added to raft group 0 and
+    transitioned to the appropriate state (`bootstrapping` or `replacing`).
```
After. This should be implied by "If the joining node accepts the decision and responds successfully".

piodul commented on service/storage_service.cc:
```diff
@@ -2148,6 +2148,45 @@ future<> storage_service::store_join_cookie() {
+    topology_mutation_builder builder(guard.write_timestamp());
+    auto& node_builder = builder.with_node(params.host_id)
+        .set("node_state", node_state::none)
+        .set("datacenter", params.datacenter)
```
Yes. Those are only verified by the node handling JOIN_NODE_REQUEST. We don't have corresponding columns for them in `system.topology`.

piodul commented:
Following our discussion on the last meeting, I'll try to simplify the implementation not to use join cookies. For now, I have responded to the comments that I was sure will be relevant after the changes.

piodul commented on idl/storage_service.idl.hh:
```diff
@@ -53,4 +53,65 @@ struct raft_topology_pull_params {};
+};
+
+struct join_node_response_result {
+    enum class response_status : uint8_t {
```
In v2, response_result is empty so I guess this comment is no longer relevant.

piodul commented on service/storage_service.cc:
```diff
@@ -1825,8 +1825,60 @@ class topology_coordinator {
+                    co_await coroutine::exception(std::make_exception_ptr(
+                            std::runtime_error(::format("no ip address mapping for {}", node.id))));
+                }
+                auto response = co_await ser::storage_service_rpc_verbs::send_join_node_response(
```
Currently, the topology coordinator will not attempt to enable features if there are pending join requests (nodes in `none` state), so there is no risk.

If we wanted to first add the node to topology and only after that send join_node_response, then we need an additional step in the state machine (because we need to know whether we may need to send the response to the joining node).

kbr-scylla replied:
I was actually hoping that we change the procedure to add to topology first, then group 0 second.

Note that when removing a node, already do it this way: we first remove it from group 0, then "remove" it from topology (actually move it to "left" state). So I think we should do the reverse when joining the node; that way, a node's lifetime in group 0 will be bounded from above by the lifetime in topology, making things easier to reason about, and also making error recovery easier e.g. after a failed boot -- the coordinator could make decisions based just on topology state.

cc @gleb-cloudius

gleb-cloudius replied:
I was distrusted from reviewing the series so did not get to here yet. but in the original proposal we added to topology before joining to the group0 IIRC.

piodul replied:
Now, the order for the accepted nodes is:

1. Move node to bootstrapping/replacing state
2. Add to group 0
3. Send join_node_response RPC

piodul commented on service/storage_service.cc:
```diff
@@ -1825,8 +1825,60 @@ class topology_coordinator {
+
+                // If the validation failed, we will remove the request at the end.
+                bool should_drop = std::holds_alternative<join_node_response_params::rejected>(decision);
+                auto ip = _address_map.find(node.id);
```
Forgot to do in v2, but will do on the next occasion (after the next round of comments).

piodul commented on db/system_keyspace.cc:
```diff
@@ -2676,8 +2676,13 @@ service::topology_features system_keyspace::decode_topology_features_state(::sha
-            ret.normal_supported_features.emplace(host_id, decode_features(deserialize_set_column(*topology(), row, "supported_features")));
+        if (row.has("supported_features")) {
+            auto features = decode_features(deserialize_set_column(*topology(), row, "supported_features"));
+            if (nstate == service::node_state::normal) {
```
Hmm, right. I think it implicitly works correctly right now, because adding nodes and transitioning nodes has more priority in the topology coordinator right now than enabling features.

There are two ways in which a node can be protected from a hazardous situation (enabling a feature when a node in the cluster does not support it):

1. Include that node when calculating the intersection of supported features of all nodes,
2. Forbid enabling features when the node is in some state

Currently, we apply 1. for normal nodes, and 2. for others (bootstrapping nodes, just-joining nodes).

I suppose we could get rid of 2. completely if we:

- Include transitioning (i.e. bootstrapping) nodes when calculating features to enable and send the read barrier request to them as well,
- Do a topology state machine transition before sending join_node_response, like I mentioned here: https://github.com/scylladb/scylladb/pull/15196#discussion_r1309819542

kbr-scylla replied:
Hm, TBH I don't see advantages of one solution or the other.
@gleb-cloudius WDYT?

But if we stay with the current approach -- i.e. the precondition of enabling is that all nodes are NORMAL -- this place should have a comment to explain why the code is safe (we don't need to interset with non-normal nodes -- they must be empty).

gleb-cloudius replied:
I think we should be able to enable features while some topology operations are still waiting to be executed in the request queue. It is fine to wait for running operation (bootstrap/replace) to complete though.

piodul replied:
Now it's safe to enable features when there are nodes in `none` state, but not when there are transitioning nodes. Adjusted the check in `enable_features` and clarified explanations in the comments why it works and considering only normal nodes for barrier and feature calculation is safe.

piodul commented on docs/dev/topology-over-raft.md:
```diff
@@ -216,3 +216,94 @@ There are also a few static columns for cluster-global properties:
+  eventually removes it.
+- If there is no topology request for that node, it is placed in
+  the `system.topology` table and an `ok` response is sent back.
+- If the node has already joined, an `ok` response is sent back.
```
Here, I meant "topology coordinator accepted it". I hope it's more clear in the current version of the doc.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3573/

avikivity commented on idl/storage_service.idl.hh:
```diff
@@ -49,7 +49,7 @@ struct raft_topology_snapshot {
-verb [[cancellable]] tablet_stream_data (locator::global_tablet_id);
+verb raft_topology_cmd (raft::server_id dst_id, raft::term_t term, uint64_t cmd_index, service::raft_topology_cmd) -> service::raft_topology_cmd_result;
+verb [[cancellable]] raft_pull_topology_snapshot (service::raft_topology_pull_params, raft::server_id dst_id [[version 5.4.0]]) -> service::raft_topology_snapshot;
+verb [[cancellable]] tablet_stream_data (raft::server_id dst_id, locator::global_tablet_id);
```
Please update the changelog to explain why changing the signature is safe.

piodul replied:
Done

avikivity commented on idl/join_node.idl.hh:
```diff
@@ -0,0 +1,58 @@
+
+struct join_node_request_result {
+    struct ok {};
+    struct accepted {};
```
What's the difference between ok and accepted? Maybe "ok" should be "already_a_member" to be clearer?

avikivity commented on service/raft/join_node.hh:
```diff
@@ -0,0 +1,65 @@
+    struct ok {};
+
+    // The request was already accepted by the topology coordinator.
+    struct accepted {};
```
Looks like my guess was exactly wrong, so "accepted" should be "already_a_member" to reduce ambiguity.

piodul replied:
Now there is no `accepted` response, only `ok` + `rejected`, so I guess it's not that ambiguous anymore.

avikivity commented:
Looks good from a shallow review. I'd like to make sure nothing is lost with upgrades in non-experimental operation.

avikivity commented:
Also, upgrade from non-experimental to consistent topology should work.

gleb-cloudius commented:
> 
>     * Snitch name is now persistently stored in `system.scylla_local` and verified on startup,

What for? Snitch name is already persistently stored in the snitch config file and there is a procedure that allows changing it.

> 
>     * Storage service verbs related to raft now verify destination host_id (without it, node replace tests would fail),
> 

Why didn't it fail before?

piodul commented:
> > ```
> > * Snitch name is now persistently stored in `system.scylla_local` and verified on startup,
> > ```
> 
> What for? Snitch name is already persistently stored in the snitch config file and there is a procedure that allows changing it.

Hm, indeed. The procedure requires the user to stop all nodes, change the snitch and only then restart the nodes.

I was confused by the fact that it's not possible to have cluster with mixed snitches (https://github.com/scylladb/scylladb/issues/6832) and wanted to disable the shadow round check, as I did with features.

I'll drop the snitch name verification on startup and will bring back the check in shadow round.

> 
> > ```
> > * Storage service verbs related to raft now verify destination host_id (without it, node replace tests would fail),
> > ```
> 
> Why didn't it fail before?

I introduced it when I was trying to fix failing `test_blocked_bootstrap`. Ultimately, I disabled that test because I believe it's not possible to make it work without failure handling.

In a nutshell, due to how the error injection is placed in that test, the joining node would stop and the topology coordinator would keep trying to bootstrap the stopped node. Then, a node replace is attempted, and the replacing node would get a bootstrap request it wasn't supposed to get - and it actually happened before group 0 was started, so on_internal_error was triggered.

I just checked and no topology test really requires it. I think it's possible to trigger such a scenario even without this PR, so I can send the relevant commit separately.

gleb-cloudius commented on service/raft/group0_state_machine.cc:
```diff
@@ -199,16 +199,16 @@ future<> group0_state_machine::load_snapshot(raft::snapshot_id id) {
 }
 
-future<> group0_state_machine::transfer_snapshot(gms::inet_address from, raft::snapshot_descriptor snp) {
+future<> group0_state_machine::transfer_snapshot(gms::inet_address from_ip, raft::server_id from_id, raft::snapshot_descriptor snp) {
```
Why do we need both ip and id? The caller translate from id to ip to call the function.

piodul replied:
I think you meant callee (`group0_state_machine::transfer_snapshot`)? Yes, I see that we can access the raft address from there and perform the translation in that function, so the `from_ip` argument is not needed.

piodul replied:
Removed the `from_ip` argument, the function now translates the host_id to IP itself. I had to pass the raft address map to `group0_state_machine` for that, though.

gleb-cloudius commented on idl/storage_service.idl.hh:
```diff
@@ -49,7 +49,7 @@ struct raft_topology_snapshot {
-verb [[cancellable]] raft_pull_topology_snapshot (service::raft_topology_pull_params) -> service::raft_topology_snapshot;
-verb [[cancellable]] tablet_stream_data (locator::global_tablet_id);
+verb raft_topology_cmd (raft::server_id dst_id, raft::term_t term, uint64_t cmd_index, service::raft_topology_cmd) -> service::raft_topology_cmd_result;
+verb [[cancellable]] raft_pull_topology_snapshot (service::raft_topology_pull_params, raft::server_id dst_id [[version 5.4.0]]) -> service::raft_topology_snapshot;
```
This verb is still behind experimental. Why versioning it?

piodul replied:
There was no need. Now it's not versioned.

gleb-cloudius commented on docs/dev/topology-over-raft.md:
```diff
@@ -218,3 +218,60 @@ There are also a few static columns for cluster-global properties:
+  coordinator's decision about the node - whether the request was accepted
+  or rejected.
+
+The procedure is designed to be idempotent. In case the joining node crashes
```
So I wonder if we should try to support the scenario where bootstrapping node restarts in the middle. I tested what happens in this case with the gossiper and it looks like the scenario is not supported. If the node restart (at least while streaming) the next bootstrap attempt will fails with an error that says that the node is already in the cluster and removenode should be used before trying the bootstrap once more. @asias?

kbr-scylla replied:
@gleb-cloudius you will probably get different result if you wait for 2 minutes before retrying the bootstrap

gleb-cloudius replied:
I know. Because the node will be flushed from the gossiper. But is this really what we want to emulate? If an admin had to know to wait 2 minutes after failure he may as well know to remove the data dir and start new bootstrap immediately.

kbr-scylla replied:
Agreed

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -6119,6 +6155,180 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
 }
 
+future<join_node_request_result> storage_service::join_node_request_handler(join_node_request_params params) {
+    if (!_raft_topology_change_enabled) {
```
May be just not register it in this case?

gleb-cloudius replied:
If not it is better to return `disabled` result instead of throw.

piodul replied:
Now the verbs are not registered unless `_raft_topology_change_enabled` is true.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -6119,6 +6155,180 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+        return !_topology_state_machine._topology.is_empty();
+    });
+
+    // TODO: Aborts?
```
Well, you already have an `_abort_source` but just decided to ignore the exception for some reason.

piodul replied:
Now the abort requested exception is propagated.

gleb-cloudius replied:
Should TODO be dropped then?

piodul replied:
Yeah, I removed it.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -6119,6 +6155,180 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+                    ignored_ids = replace->ignored_ids;
+                    ignored_ids.insert(replace->replaced_id);
+                }
+            }
```
Can ignored nodes be taken from the config params instead of through the group0?

piodul replied:
I originally wanted to do it like this, but we need `replacement_info` which is calculated in `join_token_ring` and I didn't see a way to pass it to the rpc handler in a non-awkward way (add a field to storage_service? register the rpc verb in join_token_ring and capture replacement_info with a lambda?).

gleb-cloudius replied:
Yes, good point. Fine to get them from the state machine indeed.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -2848,6 +2871,10 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
     }
 
+    if (!_group0->joined_group0()) {
+        co_await check_join_params_on_joining_node(join_params);
```
Can we check them on the recipient instead? If the node already exists in `none` state in the topology already features in the request and in the topology should match.

piodul replied:
Yes - in the previous impl where adding to group 0 was done only after finishing join_node_response it was necessary, but now I can do it solely on the join_node_request recipient.

piodul replied:
Removed, now verification happens on the join_node_request recipient.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1897,8 +1897,16 @@ class topology_coordinator {
-            case node_state::normal:
-            case node_state::none: {
+            case node_state::none:
+                if (_topo_sm._topology.normal_nodes.empty() && _topo_sm._topology.transition_nodes.empty()) {
```
Is `transition_nodes` check is needed. How can we process a node in state `none` while there is another node that is in transition?

gleb-cloudius replied:
Remove please. It confused me a little.

piodul replied:
Removed.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -2069,6 +2077,68 @@ class topology_coordinator {
+        release_guard(std::move(node.guard));
+
+        if (node_accepted && !_raft.get_configuration().contains(node.id)) {
+            co_await _raft.modify_config({raft::config_member({node.id, {}}, {})}, {});
```
@kbr-scylla so it looks like the code is already working as we discussed today on the call. When we add the node here to group0 it is already in the topology and is processed by the topology coordinator.

kbr-scylla replied:
Hm, yeah it is in `none` state, which is more like being a candidate for a bootstrapping node, not yet a really bootstrapping node.

I was thinking that we should confirm that this node can bootstrap and transition its state to `bootstrapping` or something before we add it to group 0. I.e. we accept it by doing a state machine transition, and only then add it to group 0.

In particular we'd start taking the node into account when calculating features or generally making cluster-wide decisions.

kbr-scylla replied:
But dunno if it makes any difference in practice, whether it would help failure recovery and so on.

gleb-cloudius replied:
No, you are correct. We do need to move those lines a little bit below after moving to the bootstrapping state. Theoretically it is possible that a coordinator will crash and choose another node to work on.

piodul replied:
Currently, we transition from `node_state::none` -> `node_state::bootstrapping` and also transition the topology to `commit_cdc_generation` in the same step. I'll have to separate them so that validation + transition too bootstrap is one step, and cdc generation broadcast + moving to `commit_cdc_generation` is another. Similar thing needs to be done with `replace`.

I'll also add a new state to transition the node to when we reject, so that we know we still have to deliver the rejection before definitely moving it to the `left` state.

gleb-cloudius replied:
Why should we necessary split `commit_cdc_generation` transition?

piodul replied:
Otherwise it would result in weird interleaving with some of the CDC logic.

For reference, this is how we currently handle joining nodes in `none` state:

https://github.com/scylladb/scylladb/blob/bb7121a1fbdd6b3a6be76486e44f233b91a6e2d2/service/storage_service.cc#L1971-L1994

Verification of the parameters would have to be done here, then `prepare_and_broadcast_cdc_generation_data` would be called and ultimately we would change the state to `bootstrap`/`commit_cdc_generation`. Next time the topology coordinator runs, it handles the `commit_cdc_generation` state ([here](https://github.com/scylladb/scylladb/blob/bb7121a1fbdd6b3a6be76486e44f233b91a6e2d2/service/storage_service.cc#L1715)) - I suppose we could send the join_node_response there, but it just looks out of place, and would have to be skipped if `commit_cdc_generation` was issued by the user via the `check_and_repair_cdc_streams` command.

piodul replied:
> But it looks like making modify_config() call part of handling of commit_cdc_generation will achieve the same.

I tried this approach but it looked bad to me. Before I rebased (now, tablets added some more topology transitions) the transitions actually looked like this:

- Bootstrap requests are transitioned to `node_state::bootstrapping` + `commit_cdc_generation`
- Replacing requests are transitioned to `node_state::replacing` + `write_both_read_old`

This means that we would have to take care of adding the node to group 0 during handlers for both `commit_cdc_generation` and `write_both_read_old`; what made it worse is that `commit_cdc_generation` transitions to `write_both_read_old`. We could detect whether we need to send join_node_response with strategically placed ifs but it makes the state machine flow more convoluted than it needs in my opinion.

gleb-cloudius replied:
OK. Indeed since bootstrap and replace goes through different states not introducing a common state makes things more complicated.

gleb-cloudius reviewed (commented):
Overall looks fine. I think we need to consider to not support node restart while bootstrapping. It will probably simplify the code some.

piodul commented:
v3:
- Rebased
- Now, after topology coordinator accepts a node, it is moved to `bootstrapping` or `replacing` without starting topology transition. When handling `bootstrapping`/`replacing`, node is added to group 0 and join node response is sent, then topology transition starts.
- Removed unnecessary versioning for `raft_pull_topology_snapshot`
- Handlers for `join_node_request` and `join_node_response` are not registered in non-topology-on-raft mode
- Join request parameters are now checked on the recipient
- Removed unnecessary `transition_nodes` check when topology coordinator tries to detect the first node
- Abort requested exception is propagated on the join_node_request recipient

Todos:
- Rename the structures for join node RPCs to align better with their meaning

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -2032,8 +2032,40 @@ class topology_coordinator {
+                        // Find the IP first.
+                        auto ip = get_ip_for_joining_node_response(node.id);
+
+                        // Transition to left
+                        topology_mutation_builder builder(node.guard.write_timestamp());
+                        builder.with_node(node.id)
+                               .del("topology_request")
+                               .set("node_state", node_state::left);
+                        auto reason = ::format("bootstrap: node rejected");
+                        co_await update_topology_state(std::move(node.guard), {builder.build()}, reason);
+
+                        // Perform a single attempt to tell the node that it was removed.
+                        try {
+                            co_await respond_to_joining_node(node.id, ip, join_node_response_params{
+                                .response = std::move(validation_result),
+                            });
+                        } catch (...) {
+                            slogger.warn("raft topology: failed to send rejection response to {} ({}). Will not retry.", node.id, ip);
+                        }
+
+                        slogger.info("raft topology: rejected node moved to left state {} ({})", node.id, ip);
```
I think it is fine (given that the joining node has a timeout and not just hangs). But how another coordinator can arrive to a different conclusion? It works on the same state.

piodul replied:
> But how another coordinator can arrive to a different conclusion? It works on the same state.

Right, if it works on the same state then it will arrive to the same conclusion.

A slightly different scenario then: topology coordinator sends reject response but fails to update `update_topology_state` because another node wrote a request to join. The topology coordinator retries and picks another node, bootstraps it, and only then looks at the node that was originally rejected, but the state is different. To be fair, I don't see how the state could change that the first node is rejected and then accepted, but validation might behave differently in the future.

gleb-cloudius replied:
Yes, today features can only be enabled and thus if it was not OK to add a node it will stay this way.

gleb-cloudius commented on service/raft/group0_state_machine.cc:
```diff
@@ -199,16 +200,24 @@ future<> group0_state_machine::load_snapshot(raft::snapshot_id id) {
+        // address map with its IP address.
+        const auto msg = format("Failed to apply snapshot from {}: ip address of the sender is not found", from_ip);
+        co_await coroutine::return_exception(raft::transport_error(msg));
+    }
```
It looks like there is indentation problem here.

piodul replied:
Aligned the indentation better with the surrounding code.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3774/

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -6153,6 +6189,195 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+        // We already handled this RPC. Return immediately for idempotence.
+        slogger.info("raft topology: the node got join_node_response RPC for the second time, ignoring");
+        join_node_response_result result;
+        co_return result;
```
nit: Why split to two lines instead if `co_return join_node_response_result{}`? Below as well.

piodul replied:
No good reason. Replaced them with `co_return join_node_response_result{}`

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -6235,12 +6460,25 @@ void storage_service::init_messaging_service(sharded<db::system_distributed_keys
         _messaging.local().unregister_node_ops_cmd(),
-        ser::storage_service_rpc_verbs::unregister(&_messaging.local())
+        ser::storage_service_rpc_verbs::unregister(&_messaging.local()),
+        ser::join_node_rpc_verbs::unregister(&_messaging.local())
```
Did you check that it is fine to unregister if not registered? I hope it is.

piodul replied:
Yes, I checked and it's safe. It will just be a no-op: https://github.com/scylladb/seastar/blob/576ee47d91fb48bf34801eba182d086b55c28c4f/include/seastar/rpc/rpc_impl.hh#L684

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -2216,6 +2253,59 @@ class topology_coordinator {
+        co_return retake_node(co_await start_operation(), id);
+    }
+
+    gms::inet_address get_ip_for_joining_node_response(raft::server_id id) {
```
we have ` inet_address id2ip(locator::host_id id)`

gleb-cloudius commented on db/system_keyspace.cc:
```diff
@@ -2575,10 +2575,13 @@ future<service::topology> system_keyspace::load_topology_state() {
+                    [] (auto& p) { return p.second.state != service::node_state::left_token_ring
+                                        && p.second.state != service::node_state::bootstrapping
+                                        && p.second.state != service::node_state::replacing
+                                        && p.second.state != service::node_state::rebuilding; });
```
Will it be better to add one more transition state instead? `join_group0`?

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -2497,6 +2497,92 @@ canonical_mutation storage_service::build_mutation_from_join_params(const join_n
+            // Processing of the response is done in `join_node_response_handler`.
+            // Wait for it to complete.
+            // TODO: Timeouts?
+            co_await _join_node_result.get_shared_future(_abort_source);
```
Well, since the response may get lost we need the timeout.

piodul replied:
Now there is a timeout (30s).

piodul commented on service/storage_service.cc:
```diff
@@ -2497,6 +2497,92 @@ canonical_mutation storage_service::build_mutation_from_join_params(const join_n
+        [this] (const join_node_request_result::accepted& acc) -> future<bool> {
+            // We were already added to the topology.
+            _join_node_request_done.set_value();
+
```
In the current implementation it is needed, because the joining node responds to the topology coordinator before saving the group0 ID - if there is a crash in between, it will redo the handshake but the topology coordinator will not send `join_node_response`.

I suppose the group0 handshaker interface could be mended so that the order is reversed.

I also wanted to get rid of the possibility for the nodes to retry bootstrap, as discussed [here](https://github.com/scylladb/scylladb/pull/15196#discussion_r1322872458), so `accepted` will be removed anyway. However, saving group0 ID first might still be a good idea, otherwise a node could get added by the topology coordinator, crash and then retry bootstrap, which will fail.

gleb-cloudius replied:
> In the current implementation it is needed, because the joining node responds to the topology coordinator before saving the group0 ID - if there is a crash in between, it will redo the handshake but the topology coordinator will not send join_node_response.

So we send `ok` if there is no entry in the topology or the entry is in the request state. But what if an entry is in bootstrapping state but  `join_node_response` was still not issued? As far as I see it may happen since the coordinator first move the node to `bootstrapping` state, then adds it the config and only then issues the response call.

> I also wanted to get rid of the possibility for the nodes to retry bootstrap

OK. I think it should be done before merging then.

piodul replied:
Now the procedure is not retryable, and we store the group 0 ID before ending the ` join_node_response` RPC handler.

piodul commented:
v4:
- Now, if a node fails in the middle of the join handshake, it will not be allowed to retry the handshake. In order to retry adding a node in this situation, the data directory must be cleared first.
- The group0 id is persisted first before joining node finishes its `join_node_response` RPC.
- The `join_group0` state is introduced which sits in the middle between `none` and `bootstrap`/`replace`.
- The check at the beginning of `enable_features` is adjusted, as well as the comments to explain why it is safe to do the things that it does.
- Checking whether the replaced node is in `normal` state is now done on the topology coordinator.
- Indentation adjustments, small cleanups.

piodul commented:
@avikivity

> Looks good from a shallow review. I'd like to make sure nothing is lost with upgrades in non-experimental operation.

Could you clarify what you would like me to do here? The non-experimental code path should work exactly the same as before.

> Also, upgrade from non-experimental to consistent topology should work.

This is going to be covered in https://github.com/scylladb/scylladb/issues/15008 which depends on this PR.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3805/

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -2234,6 +2267,59 @@ class topology_coordinator {
+        const auto& supported_features = node.rs->supported_features;
+        std::ranges::set_difference(node.topology->enabled_features, supported_features, std::back_inserter(unsupported_features));
+        if (!unsupported_features.empty()) {
+            slogger.info("raft topology: node {} does not understand some features: {}", node.id, unsupported_features);
```
Maybe even `warn` would be appropriate, the admin is trying to do something fishy

piodul replied:
Done, I also added more `warn`-level logs for other situations where the node is rejected

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -2037,8 +2037,40 @@ class topology_coordinator {
+                    auto validation_result = validate_joining_node(node);
+                    if (auto* reject = std::get_if<join_node_response_params::rejected>(&validation_result)) {
+                        // Find the IP first.
+                        auto ip = id2ip(locator::host_id(node.id.uuid()));
```
BTW. `id2ip` is a temporary helper, it retries forever. Once we migrate to host IDs for identification, any translation attempts we do should only happen for the purposes of messaging (like here) and should be protected with timeouts. I would thus refrain from using `id2ip`, keep using it only in these temporary use cases and not in cases like this one.

piodul replied:
> Why do it first? Should failure to do the translation result in failure to reject node boot?

In tests, I observed that this IP is usually unavailable at the beginning and it usually takes 2 attempts for the gossiped IP to reach the topology coordinator. I do the translation before transition to increase the chance of success of the `join_node_response` rpc that follows `update_topology_state`. Though, in all fairness, I should probably use `_gossiper.wait_alive` for that purpose.

> BTW. id2ip is a temporary helper, it retries forever. Once we migrate to host IDs for identification, any translation attempts we do should only happen for the purposes of messaging (like here) and should be protected with timeouts.

Ok, in that case it's another argument in favor of `_gossiper.wait_alive` before `update_topology_state`. Alternatively, I can add another state that marks a node for rejection. See my comment here: https://github.com/scylladb/scylladb/pull/15196#discussion_r1329841402

gleb-cloudius replied:
> BTW. `id2ip` is a temporary helper, it retries forever. Once we migrate to host IDs for identification, any translation attempts we do should only happen for the purposes of messaging (like here) and should be protected with timeouts. I would thus refrain from using `id2ip`, keep using it only in these temporary use cases and not in cases like this one.

Messaging should be done using ids as well, so the use of `id2ip` here is temporary as well.

gleb-cloudius replied:
> In fact, failure translation could easily happen if the bootstrapping node dies and some time passes before we reach this step.

But we decided in such a case the bootstrap will not restart, so it is fine to fail it.

gleb-cloudius replied:
> In tests, I observed that this IP is usually unavailable at the beginning and it usually takes 2 attempts for the gossiped IP to reach the topology coordinator. 

We should not rely on retry to wait for gossiper to converge any more. This will just complicate error handling (should you fail or retry?).

piodul replied:
Now, `id2ip` is put just before the RPC call and there is a retry loop with a timeout that tries to send the response.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -2234,6 +2267,59 @@ class topology_coordinator {
+            co_await _raft.modify_config({raft::config_member({id, {}}, {})}, {});
+        }
+
+        auto ip = id2ip(locator::host_id(node.id.uuid()));
```
See my other comments regarding id2ip.

avikivity commented:
> @avikivity
> 
> > Looks good from a shallow review. I'd like to make sure nothing is lost with upgrades in non-experimental operation.
> 
> Could you clarify what you would like me to do here? The non-experimental code path should work exactly the same as before.

Start with a non experimental cluster, enable consistent topology (the goal is to be sure upgrades work after it graduates from experimental mode). Should also survive downgrades.

> 
> > Also, upgrade from non-experimental to consistent topology should work.
> 
> This is going to be covered in https://github.com/scylladb/scylladb/issues/15008 which depends on this PR.

piodul commented on db/system_keyspace.cc:
```diff
@@ -2627,9 +2632,11 @@ future<service::topology> system_keyspace::load_topology_state() {
         }
-    }
 
-    ret.features = decode_topology_features_state(std::move(rs));
```
It's still used for the feature check that happens on startup see `system_keyspace::load_topology_features_state`, which is in turn used in `feature_service::enable_features_on_startup`.

piodul commented on service/storage_service.cc:
```diff
@@ -6153,6 +6189,165 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+            // This condition might not be true yet as this information is
+            // propagated through gossip. In order to reduce the chance of
+            // repair/streaming failure, wait here until we see normal nodes
+            // as UP (or the timeout elapses).
```
Yes. Updated the PR cover letter so that the issue will be "Fixed" by the PR.

piodul commented on docs/dev/topology-over-raft.md:
```diff
@@ -224,3 +224,62 @@ There are also a few static columns for cluster-global properties:
+at any point, it can restart and initiate the handshake again, but some of the
+configuration that it attempted to join with for the first time must
+be unchanged, or at least changed in a safe way (i.e. node must support
+the same features as it did for the first time, or more).
```
Hmm, I was sure I did this update to the document today. I must have lost it during a rebase. I'll try to find it or redo it again.

piodul replied:
Description of the procedure should be fixed now.

piodul commented on service/raft/raft_group0.cc:
```diff
@@ -500,6 +497,10 @@ future<> raft_group0::join_group0(std::vector<gms::inet_address> seeds, bool as_
     _group0 = group0_id;
 
+    if (!is_discovery_leader) {
+        handshaker->on_success();
```
The `perform_handshake` method is not used by the discovery leader. It's not immediately visible from the implementation (and definitely not guarded by an `if` like `on_success` call), but it's true.

As for `on_success`, it's `Called after a successful handshake`, and a handshake only happens if the node is not the discovery leader.

I can add comments to `join_group0` if this is confusing.

piodul replied:
`join_group0` method was changed once again, so I guess this is no longer applicable.

piodul commented on service/storage_service.cc:
```diff
@@ -6153,6 +6189,165 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+                if (ignored_ids.contains(id)) {
+                    continue;
+                }
+                if (auto addr = amap.find(id)) {
```
Yes, we definitely shouldn't skip. I'll update.

piodul replied:
Done

piodul commented on service/storage_service.cc:
```diff
@@ -6153,6 +6189,165 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+        } catch (abort_requested_exception&) {
+            throw;
+        } catch (...) {
+            slogger.warn("raft topology: join_node_request: failed to start group0 operation: {}, retrying.",
```
I had the following failure in `test_replace.not_sure_which_test_case`:

- Cluster with nodes `A`, `B`, `C` - `A` is the topology coordinator/group0 leader,
- `A` is killed,
- New node `D` attempts to replace `A` with the same IP as `A`, sends `join_node_request` rpc to node `B`,
- Node `B` handles the RPC and wants to perform group0 operation and wants to perform a barrier - still thinks that `A` is the leader and is alive, sends an RPC to its IP,
- `D` accidentally receives the request that was meant to `A` but throws an exception because of host_id mismatch,
- Failure is propagated back to `B`, and then to `A` - and `A` fails the replace operation.

Adding the retries makes the logic resilient to this. If you have another idea how to avoid this error, I'm all ears (eyes?).

kbr-scylla replied:
Wow, interesting sequence of events.

I think retrying is OK, but unlike `group0_concurrent_modification`, which always implies progress, I don't think we should retry `start_operation` failure forever. Maybe a timeout is appropriate here.

Also what was the type of exception that you got in this scenario? Maybe we should only catch that one instead of `...`.

piodul replied:
```
 WARN  2023-09-22 08:28:37,920 [shard 0:goss] storage_service - raft topology: join_node_request: failed to start group0 operation: seastar::rpc::remote_verb_error (Got message for server 39e64975-cde9-4c17-9601-d82913046ddf, but my id is 5b46ceba-346d-4cf0-98b1-da424f268d23), retrying.
 ```

It's `seastar::rpc::remote_verb_error`.

> I think retrying is OK, but unlike `group0_concurrent_modification`, which always implies progress, I don't think we should retry `start_operation` failure forever. Maybe a timeout is appropriate here.

I can add a timeout, though I wonder if there is a better way.

kbr-scylla replied:
@gleb-cloudius maybe you have some better idea?

gleb-cloudius replied:
Scenario is indeed weird. Worth keeping in mind. Ideas? Well here is one: `join_node_request` checks if the id of the node being replaced is currently considered to be a leader wait Nsecs for leader change. Otherwise return an error.

piodul replied:
If ID of the leader == ID of the replaced node, then it's either our case or an operator error (trying to replace a node that is alive). In the latter case, we will delay the error by N seconds (according to Kamil node reelection usually takes ~ 3s, so N >> 3?) - which is not terrible, though maybe can be avoided.

Instead of looking at IDs, maybe we could compare IPs? If the IP of the _new_ node is the same as IP of the perceived raft leader, then we could wait for a leader change.

gleb-cloudius replied:
We want to get rid of id to ip translation in upper layer. I think waiting a couple of seconds before failing is fine.

piodul replied:
Ack, I'm now comparing IDs and waiting until the leader changes.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -6153,6 +6189,165 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+        // We already handled this RPC. Return immediately for idempotence.
+        slogger.info("raft topology: the node got join_node_response RPC for the second time, ignoring");
+        co_return join_node_response_result{};
+    }
```
When it is possible to get more then one RPC? If a leader switches while handling the 'join` request? Worth explaining it in the comment. May be before the semaphore above.

piodul replied:
Added a comment at the beginning of the handler to explain why we may receive multiple RPCs (and why those can happen in parallel to each other).

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -6153,6 +6189,165 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+
+            // Delay response until we are sure that this node will not retry
+            // the handshake on next boot.
+            co_await _join_node_group0_id_persisted.get_shared_future(_abort_source);
```
Is this needed? What will happen without? The node will restart and fail to join because handshake will fail? But it is only if it restarts during bootstrap, no? And we decided the node will fail in such a case anyway,

piodul replied:
> Is this needed? What will happen without? The node will restart and fail to join because handshake will fail? But it is only if it restarts during bootstrap, no?

Yes, I agree that nothing bad should happen without it. Perhaps it's not worth complicating the control flow, so I can remove it.

piodul replied:
Actually, one bad thing could happen: the joining node, on restart, could receive a command to bootstrap itself before it sends `join_node_request` and receives an error. But that's most likely a different problem with a different fix, as the joining node should not accept such commands before a successful `join_node_request` (and we should start group0 server after that, too).

piodul replied:
Got rid of `_join_node_group0_id_persisted` promise, now server is started only after `join_node_request` finishes successfully, so the issue I described should no longer be relevant.

gleb-cloudius replied:
> Got rid of `_join_node_group0_id_persisted` promise, now server is started only after `join_node_request` finishes successfully, so the issue I described should no longer be relevant.

But I see it is still here.

gleb-cloudius replied:
But I do not see it if I checkout the pull request locally. This github thing is broken.

gleb-cloudius replied:
Ah. This is because the line is removed by some later commit. But it means this commit cannot be compiled. Please check that each individual commit compiles.

piodul replied:
Done, also made sure that all commits compile in dev mode.

gleb-cloudius commented on service/topology_state_machine.hh:
```diff
@@ -28,6 +28,7 @@ namespace service {
 
 enum class node_state: uint16_t {
     none,                // the new node joined group0 but did not bootstraped yet (has no tokens and data to serve)
+    join_group0,         // the node was accepted but it needs to be added to group 0 before bootstrap/replace
```
I suggested to add transition state, not a node state. I think it is more appropriate. We do not loose the information about what operation is ongoing.

piodul replied:
Sorry, I confused those two types of states.

@kbr-scylla WDYT? I remember asking you some time ago and you preferred a node state.

kbr-scylla replied:
I did, but I don't have any good arguments other than what you said: transition states are more for operations that change the entire cluster, while node states are for changing single node. OTOH you could say that the act of adding a new node has a significant effect, it affects feature calculation and so on, so maybe transition is indeed more appropriate.

 @gleb-cloudius why do you prefer transition state?

kbr-scylla replied:
Fine by me.

piodul replied:
Done, now `join_group0` is a transition state instead of node state.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -919,6 +921,9 @@ class topology_coordinator {
         if (topo.transition_nodes.size() != 0) {
             // If there is a node that is the middle of topology operation continue with it
             e = &*topo.transition_nodes.begin();
+            if (e->second.state == node_state::join_group0) {
```
If `join_group0` would be a transition state this will not be needed.

kbr-scylla commented:
@avikivity 

> Start with a non experimental cluster, enable consistent topology (the goal is to be sure upgrades work after it graduates from experimental mode).

Out of the scope of this PR. Implementing upgrade procedure to consistent topology is a whole other multi-week task. There's a separate issue for this: https://github.com/scylladb/scylladb/issues/15008

Recall that for raft-schema we also implemented upgrade procedure once almost everything else was finished.

piodul commented on service/raft/raft_group0.hh:
```diff
@@ -283,6 +310,10 @@ private:
+    // If the server becomes the discovery leader, meaning that is it elected as the server
+    // which creates group 0, then it will become a voter. Otherwise it will use the `handshaker`
+    // to ask the discovery leader to join the node to the group, and the voting status
+    // depends on how exactly this is implemented.
```
Done

piodul commented on service/storage_service.cc:
```diff
@@ -6153,6 +6189,165 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+                    .reason = "The node has already been removed from the cluster",
+                };
+            } else {
+                result.result = join_node_request_result::rejected{
```
Yes, updated.

piodul commented on service/storage_service.cc:
```diff
@@ -6153,6 +6189,165 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+    // Serialize handling the responses
+    auto lock = co_await get_units(_join_node_response_handler_mutex, 1);
+
+    // Wait until the join_node_request is sent.
```
Changed to something similar (`Wait until we sent and completed the join_node_request RPC`)

piodul commented on service/storage_service.cc:
```diff
@@ -6153,6 +6189,165 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+            slogger.info("raft topology: coordinator accepted request to join, "
+                    "waiting for nodes {} to be alive before responding and continuing",
+                    sync_nodes);
+            co_await _gossiper.wait_alive(sync_nodes, std::chrono::seconds(30));
```
Done

piodul commented:
v5:
- Updated description of the procedure in the `topology-over-raft.md` doc,
- Changed the `join_group0` node state to be a transition state,
- Now, group0 server is started only after `join_node_request` completes successfully,
- Got rid of `_join_node_group0_id_persisted` promise,
- The group0_handshaker was adjusted: now it has two "phases", one that happens before group0 server start and the other that happens after that,
- Updated the comment for `join_group0`,
- Now, that code that waits until normal nodes are alive resolves all of their addresses in the raft address map (with a timeout) instead of ignoring the unknown ones,
- The `join_node_request` RPC handler now tries to detect when there is an attempt to replace a leader that has just died and waits for a leadership change. The exact scenario is described in a comment,
- More detailed error messages on the node handling `join_node_request` about the reason for rejection,
- Explained why `join_node_response` can be received more than once, possibly at the same time,
- `id2ip` is now used just before issuing RPCs,

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3904/

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -6155,6 +6191,264 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+            while (!g0_server.current_leader() || *params.replaced_id == g0_server.current_leader()) {
+                // FIXME: Wait for the next term instead of sleeping in a loop
+                // Waiting for state change is not enough because a new leader
+                // might be chosen without us going through the candidate state.
```
@gleb-cloudius perhaps we should consider becoming a follower for another leader to be a state change as well, i.e. removing this `holds_alternative` `if`:
```cpp
void fsm::become_follower(server_id leader) {
    if (leader == _my_id) {
        on_internal_error(logger, "fsm cannot become a follower of itself");
    }
    if (!std::holds_alternative<follower>(_state)) {
        _output.state_changed = true;
```

kbr-scylla replied:
We can do this in a follow-up

gleb-cloudius replied:
> @gleb-cloudius perhaps we should consider becoming a follower for another leader to be a state change as well, i.e. removing this `holds_alternative` `if`:
> 

Was thinking exactly the same.

piodul replied:
Great, I'll send a followup after this PR is merged.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -6155,6 +6191,264 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+        // is going to replace but the ID of the replaced node is the same
+        // as the leader, wait for a short while until a reelection happens.
+        // If replaced ID == leader ID, then this indicates either the situation
+        // above or an operator error (actually trying to replace a live node).
```
Not a big fan of this solution -- after doing the wait for the condition to become false, the condition can get back to be true -- but if it's necessary for passing the tests, it's fine for now. We can think of some better alternative as a follow-up.

gleb-cloudius replied:
I the condition is going back to true it means that `A` is still alive and hence `D` cannot have the same IP, so the request will be added and then rejected by the coordinator if by the time it will process the request `A` will still be alive.

piodul commented:
v5.1: added more details in `topology-over-raft.md` about the consequences of starting group 0 before receiving an answer to `JOIN_NODE_REQUEST` RPC

gleb-cloudius commented on docs/dev/topology-over-raft.md:
```diff
@@ -224,3 +224,64 @@ There are also a few static columns for cluster-global properties:
+- Reject the request. This can happen if:
+  - The request does not satisfy some validity checks done by the receiving node
+    (e.g. cluster name does not match),
+  - There already an existing request to join this node (which can happen if
```
There *is*....

gleb-cloudius commented on docs/dev/topology-over-raft.md:
```diff
@@ -224,3 +224,64 @@ There are also a few static columns for cluster-global properties:
+
+If the node didn't join group 0, it sends `JOIN_NODE_REQUEST` to any existing
+node in the cluster. The receiving node can either:
+- Accept the request and tell the new node to wait for `JOIN_NODE_RESPONSE`.
```
`JOIN_NODE_RESPONSE` or `JOIN_NODE_REPLY`?

gleb-cloudius commented on docs/dev/topology-over-raft.md:
```diff
@@ -224,3 +224,64 @@ There are also a few static columns for cluster-global properties:
+  - The node was removed from the cluster.
+
+The new node should not process `JOIN_NODE_RESPONSE` RPC until it is explicitly
+ordered by node that handles `JOIN_NODE_REQUEST`. This is necessary so that
```
But cam this happen after we made the procedure not restartable?

gleb-cloudius commented on docs/dev/topology-over-raft.md:
```diff
@@ -224,3 +224,64 @@ There are also a few static columns for cluster-global properties:
+
+The new node should not process `JOIN_NODE_RESPONSE` RPC until it is explicitly
+ordered by node that handles `JOIN_NODE_REQUEST`. This is necessary so that
+it doesn't accidentally process a join request from a previous attempt, if there
```
But can this happen after we made the procedure not restartable?

piodul replied:
Yes, the new node uses `JOIN_NODE_REQUEST` to check whether it previously tried to join - the node that handles the request will reject it in such case.

I suppose the joining node could store a persistent flag which tells whether it successfully finished joining or not. I don't think it would simplify the implementation much (only that group0 handshaker would have one method instead of two) but perhaps would simplify reasoning about the procedure and this paragraph would be unnecessary.

gleb-cloudius replied:
I try to remove flags, not add them If the situation may indeed happen I am fine with the way it is handled.

gleb-cloudius commented on docs/dev/topology-over-raft.md:
```diff
@@ -224,3 +224,64 @@ There are also a few static columns for cluster-global properties:
+ordered by node that handles `JOIN_NODE_REQUEST`. This is necessary so that
+it doesn't accidentally process a join request from a previous attempt, if there
+was any. Similarly, the group 0 server must not be started on the new node
+until the `JOIN_NODE_REQUEST` RPC succeeds, otherwise it might receive some
```
`JOIN_NODE_REQUEST` or `JOIN_NODE_RESPONSE`? Only after coordinator decided that it is safe can we start group0, no?

piodul replied:
> `JOIN_NODE_REQUEST` or `JOIN_NODE_RESPONSE`?

`JOIN_NODE_REQUEST`.

> Only after coordinator decided that it is safe can we start group0, no?

The coordinator will either:

- Decide to accept us, add us to the configuration and will send `JOIN_NODE_RESPONSE` to us that we are accepted,
- Decide to reject us, _not_ add us to the configuration and will send `JOIN_NODE_RESPONSE` to us that we are rejected.

Therefore we will only receive group0 commands if we get accepted.

I suppose we could start group0 server when we receive `JOIN_NODE_RESPONSE`, but we are added to group0 configuration by the coordinator before it sends `JOIN_NODE_RESPONSE`, so some commands might be rejected because we didn't start group0 server yet.

gleb-cloudius replied:
Command will be retired. The logic above makes sense though. I wanted just make sure the the doc is accurate. You can even extend it with the explanation above why it is safe.

But according to the same logic it is safe to create group0 server even before `JOIN_NODE_REQUEST` succeeds, no?

kbr-scylla replied:
> But according to the same logic it is safe to create group0 server even before JOIN_NODE_REQUEST succeeds, no?

I had the same question, we discussed this with Piotr D on slack :D

IIUC, this server might have been added to group 0 configuration on the previous join attempt. Now, if we started the server before JOIN_NODE_REQUEST, we'd start receiving commands, which would be incorrect (we might have restarted with a smaller set of features so we don't understand some commands). But if we start only after successful JOIN_NODE_REQUEST, in the case of second attempt, we would get rejected and never start the server, so we're protected from commands that we may not understand.

gleb-cloudius commented on docs/dev/topology-over-raft.md:
```diff
@@ -224,3 +224,64 @@ There are also a few static columns for cluster-global properties:
+  an immediate failure of the RPC -  but others need to be verified
+  by the topology coordinator.
+
+- `JOIN_NODE_REPLY`
```
In the code and in the rest of the doc the rpc is called `JOIN_NODE_RESPONSE`

gleb-cloudius commented on docs/dev/topology-over-raft.md:
```diff
@@ -264,10 +264,11 @@ node in the cluster. The receiving node can either:
-was any. Similarly, the group 0 server must not be started on the new node
-until the `JOIN_NODE_REQUEST` RPC succeeds, otherwise it might receive some
-commands it does not understand due to unsupported features.
+it doesn't accidentally process a join response from a previous attempt, if
+there was any. Similarly, the group 0 server must not be started on the new node
+until the `JOIN_NODE_REQUEST` RPC succeeds, otherwise, if it crashes during
+the prodecude and disables some features it might receive some commands it is
+not prepared to understand.
```
Should this be squashed with the documentation patch?

piodul replied:
I squashed my last fix to `topology-over-raft.md` with a wrong commit. Will fix.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -6155,6 +6191,264 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+        // - `D` accidentally receives the request that was meant to `A`
+        //   but throws an exception because of host_id mismatch,
+        // - Failure is propagated back to `B`, and then to `A` - and `A`
+        //   fails the replace operation.
```
In the last step `A` should be replaced with `D`.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -6155,6 +6191,264 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+        t.set_callback([&as] {
+            as.request_abort();
+        });
+        t.arm(timeout);
```
There is `with_timeout` in `seastar/core/with_timeout.hh`

piodul replied:
Personally, I'm not a big fan of `seastar::with_timeout` because it doesn't properly cancel the future that is being waited on. I prefer telling fibers to abort and then waiting until they gracefully finish.

gleb-cloudius replied:
You can pass abort source to the function you execute under `with_timeout` and wait through a gate. It may still be simpler than to re-invent it here. But I agree that `with_timeout` should have bee designed to return original future in case of timeout.

piodul replied:
I'll address it in the followup, I'll be changing the logic here anyway (according to Kamil's yesterday comments).

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1784,6 +1786,47 @@ class topology_coordinator {
+                    case node_state::bootstrapping: {
+                        assert(node.rs->ring);
+                        auto tmptr = get_token_metadata_ptr();
+                        auto [gen_uuid, guard, mutation] = co_await prepare_and_broadcast_cdc_generation_data(
```
Can you explain why you decided to move CDC here, but not token assignment? I think both can stay where they are now.

piodul replied:
We have an existing assertion in `load_topology_state` which checks that transitioning nodes (except for `left_token_ring` state) have tokens assigned, so I kept the logic that assigns the token in the first step.

As for the CDC part, I didn't want `prepare_and_broadcast_cdc_generation_data` to be called while a handshake is in progress. I wanted the handshake to take as little time as possible to reduce the chance of failure, and `prepare_and_broadcast_cdc_generation_data` may potentially transfer a large chunk of data.

gleb-cloudius replied:
I see.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -2238,6 +2278,59 @@ class topology_coordinator {
+                    .reason = ::format("Cannot replace node {} because it is not in the 'normal' state", replaced_id),
+                };
+            }
+        }
```
We can also check that it is dead here. But in a follow up.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -2077,8 +2078,47 @@ class topology_coordinator {
+
+                        // Keep trying to send the rejection to the node, give up after some time.
+                        const auto send_reject_deadline = lowres_clock::now() + std::chrono::seconds(30);
+                        while (true) {
```
Can this be done outside of a state machine? The state machine can proceed doing other topology operations while we try to kill a node. Also if coordinator changes after the node is marked left we will not try to shutdown. May be we should move to `node_state::left_token_ring` and handle the shutdown there they way we handle it during decommission: try one, if failed print a message that the node can be shutdown manually.

kbr-scylla replied:
It makes sense, but can we do it in a follow-up?

gleb-cloudius replied:
We can.

piodul commented:
v5.2:
- Fixes to the documentation and comments
- Fixups misapplied in the previous version should now be applied correctly
- Tiny cleanup in the `transition_state::join_group0` handler
- Made sure that all commits compile in dev mode

Diff should be available here: https://github.com/scylladb/scylladb/compare/97b30a1ec5dae31baedfac63b3ce19ff304cf90d..d3c29d481c7bf5f76f4c983d98128c6b544e4ee7

piodul commented:
Will rebase in a second.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3939/

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -11,6 +11,8 @@
 #include "storage_service.hh"
 #include "dht/boot_strapper.hh"
+#include <boost/range/adaptor/filtered.hpp>
+#include <boost/range/adaptor/transformed.hpp>
```
Nit: I think those are no longer needed, no?

piodul replied:
Indeed, I checked and it's no longer needed. Will remove in the followup.

piodul commented:
v5.4:
- Worked around an issue described in https://github.com/scylladb/scylladb/issues/15555#issuecomment-1737484966 which appears to fail in our CI but didn't occur locally (join_node_response handler waits for group0 to be started before continuing)
- Removed the unnecessary `boost::range` imports as a bonus

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3956/

piodul commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3956/

Apparently, we exceeded some rate limit in GitHub and that caused the CI to fail. Rerunning.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3961/
